backend/packages/Upgrade/src/api/services/MoocletRewardsService.ts

```diff
@@ -0,0 +1,207 @@
+import { InjectRepository } from '../../typeorm-typedi-extensions';
+import { BinaryRewardMetricAllowedValue, BinaryRewardMetricValueMap } from '../../../../../../types/src/Mooclet';
+
+interface FindRewardParams {
+  user: RequestedExperimentUser;
+  moocletExperimentRefs: MoocletExperimentRef[];
+  logger: UpgradeLogger;
+}
+
+interface SyncRewardParams {
+  moocletExperimentRef: MoocletExperimentRef;
+  metricValue: string | number;
+  logger: UpgradeLogger;
+}
+
+@Service()
+export class MoocletRewardsService {
+  constructor(
+    @InjectRepository()
+    private moocletExperimentRefRepository: MoocletExperimentRefRepository,
+    @InjectRepository()
+    private individualEnrollmentRepository: IndividualEnrollmentRepository,
+    private metricService: MetricService,
+    private moocletDataService: MoocletDataService
+  ) {}
+
+  public attachRewardMetricQuery(rewardMetricKey: string, queries: any[]) {
+    queries.push({
+      id: uuid(),
+      name: 'Success Rate',
+      query: {
+        operationType: 'percentage',
+        compareFn: '=',
+        compareValue: BinaryRewardMetricAllowedValue.SUCCESS,
+      },
+      metric: {
+        key: rewardMetricKey,
+      },
+      repeatedMeasure: REPEATED_MEASURE.mostRecent,
+    });
+  }
+
+  public async createAndSaveRewardMetric(
+    rewardMetricKey: string,
+    context: string,
+    logger: UpgradeLogger
+  ): Promise<Metric[]> {
+    const metric = {
+      id: uuid(),
+      metric: rewardMetricKey,
+      datatype: IMetricMetaData.CATEGORICAL,
+      allowedValues: [BinaryRewardMetricAllowedValue.SUCCESS, BinaryRewardMetricAllowedValue.FAILURE],
+    };
+
+    return this.metricService.saveAllMetrics([metric], [context], logger);
+  }
+
+  public async parseLogsForActiveRewardMetricKey(
+    user: RequestedExperimentUser,
+    logs: ILogInput[],
+    logger: UpgradeLogger
+  ): Promise<void> {
+    try {
+      const moocletExperimentRefs = await this.moocletExperimentRefRepository.getRefsForActivelyEnrollingExperiments();
+
+      if (!moocletExperimentRefs.length) {
+        logger.info({
+          message: 'No active mooclet experiment refs found',
+          user,
+        });
+        return;
+      }
+
+      const searchParams: FindRewardParams = { user, moocletExperimentRefs, logger };
+
+      this.searchLogAttributes(logs, searchParams);
+    } catch (error) {
+      logger.error({ message: 'Failed to fetch active mooclet refs ', error });
+    }
+  }
+
+  private searchLogAttributes(logs: ILogInput[], params: FindRewardParams): void {
+    logs.forEach((log) => {
+      const {
+        metrics: { attributes },
+      } = log;
+
+      const attributesKeys = Object.keys(attributes);
+      for (const key of attributesKeys) {
+        this.findExperimentByRewardMetricKey(key, attributes[key], params);
+      }
+    });
+  }
+
+  private findExperimentByRewardMetricKey(key: string, value: string | number, searchParams: FindRewardParams): void {
```

Comment:
There's no guarantee that `value` here will be either 'SUCCESS' or 'FAILURE'. when, for any reason, /log is called with something else, a reward variable is created with no value. That's probably fine if we're sure those are ignored w.r.t updating the algorithm. But if we're sending data that'll just be ignored (or maybe won't), it could save traffic to just not send it in those cases (or error out).

Reply:
true, we should log a warning about a bad value and not send anything.